lib/pf/services/manager/tc.pm

```diff
@@ -0,0 +1,312 @@
+    if (-f "$install_dir/var/tc_applied") {
+        return $TRUE;
+    } else {
+        return $FALSE;
+    }
+}
+
+=head2 pid
+
+Override the default method to check pid since there really is no such thing for tc (it's not a process).
+
+=cut
+
+sub pid {
+    my $self   = shift;
+    my $result = `sudo systemctl show -p ActiveState packetfence-tc`;
+    chomp $result;
+    my $state = ( split( '=', $result ) )[1];
+    if ( grep { $state eq $_ } qw( active activating deactivating ) ) {
+        return -1;
+    }
+    else { return 0; }
+}
+
+=head2 manageTrafficShaping
+
+Add or remove traffic shaping on the system
+
+=cut
+
+sub manageTrafficShaping {
+    my $add_tc = @_;
+    my $logger = get_logger();
+
+    if (!$add_tc) {
+        if (-f "$install_dir/var/traffic_shaping.bak") {
+            open (my $fh, "$install_dir/var/traffic_shaping.bak");
+            while (my $row = <$fh>) {
+                chomp $row;
+                my $cmd = untaint_chain($row);
+                my @out = pf_run($cmd);
+            }
+            close $fh;
+        }
+        if (-f "$install_dir/var/tc_applied") {
+           unlink("$install_dir/var/tc_applied");
+        }
+    } else {
+        open (my $fh, "+>$install_dir/var/traffic_shaping.bak");
+        my $iptables = pf::iptables->new();
+        my $indice = 100;
+        my $index = {};
+        my $indice2 = 1;
+        my $cmd;
+        my @out;
+
+        my $full_path = can_run('tc');
+        open (my $fa, "+>$install_dir/var/traffic_shaping.add");
+
+        my @ints = split(',', $iptables->get_network_snat_interface());
+        push @ints,  $management_network->tag("int");
+        foreach my $int (@ints) {
+            $index->{$int} = $indice;
+            $indice --;
+        }
+
+        foreach my $interface (@internal_nets) {
+            my $dev = $interface->tag("int");
+            my $enforcement_type = $Config{"interface $dev"}{'enforcement'};
+            if (is_type_inline($enforcement_type)) {
+                $index->{$dev} = $indice2;
+                $indice2 ++;
+            }
+        }
+
+        foreach my $key (keys %$index) {
+            $cmd = "sudo $full_path qdisc add dev $key root handle $index->{$key}:0 htb default 1";
+            print $fa $cmd."\n";
+            $cmd = untaint_chain($cmd);
+            @out = pf_run($cmd);
+            my $cmd_remove = "sudo $full_path qdisc del dev $key root";
+            print $fh $cmd_remove."\n";
+        }
+
+        my @roles = pf::nodecategory::nodecategory_view_all;
+
+        foreach my $network ( keys %NetworkConfig ) {
+
+            next if ( !pf::config::is_network_type_inline($network) );
+            my $dev = $NetworkConfig{$network}{'interface'}{'int'};
+
+            my $gateway = (defined $NetworkConfig{$network}{'next_hop'} ? $NetworkConfig{$network}{'next_hop'} : $NetworkConfig{$network}{'gateway'});
+
+            my @interface_src = split(" ", pf_run("sudo ip route get 8.8.8.8 from $gateway"));
```

Comment:
This code looks familiar!

Couldn't this be put in a util function so that both code paths use the same code and we only have a single place to adjust it in the the output of that command changes